README.md
# uxcore-calendar

---

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url]
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-calendar.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-calendar
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-calendar.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-calendar
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-calendar.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-calendar?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-calendar.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-calendar
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-calendar.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-calendar#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-calendar.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-calendar.svg
[sauce-url]: https://saucelabs.com/u/uxcore-calendar

## TL;DR

transfer ui component for react

#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-calendar
$ cd uxcore-calendar
$ npm install
$ npm start
```

## Usage

```js
var Calendar = require('uxcore-calendar');
const {
  MonthCalendar, YearCalendar, RangeCalendar, CalendarFull, MiniWeek
} = Calendar;
React.render(
  (<Calendar />),
  document.getElementById('target')
);
```

### demo
http://uxco.re/components/calendar/

## API
- onSelect(date, formatDateString)
    - date `date`
    - formatDateString `string`

## Attention

0.8.0 之后，我们使用 moment 替代了 gregorianCalendar，在一些格式上和用法上有了一些变化。


## 共通 Props

|参数|说明|类型|默认值|
|---|----|---|------|
|placeholder|占位符|string|请选择日期|
|locale|`en-us`,`zh-cn` 和 `pl-pl`|string|`zh-cn`|
|disabledDate|不可选择的日期|(currentDate) => boolean|无|
|onSelect|当日期被选中时触发|function(date)|无|
|disabled|禁用|boolean|false|
|getPopupContainer| 弹出的菜单渲染在哪个容器中 | function(trigger:Node):Node | () => document.body|
|size| 尺寸，支持 large/middle/small | string | large |
|allowClear| 是否支持清空 | boolean | true |
|onOpenChange | 在下拉面板展开、收起时调用 | function(open) | noop |
|defaultOpenValue| 在值为空的情况下，日历面板默认展示的值 | boolean |  |

### props

|参数|说明|类型|默认值|
|---|----|---|------|
|value|日期|string/number|无|
|defaultValue|日期|string/number|无|
|format|展示的日期格式|string|'YYYY-MM-DD'|
|disabledTime|不可选择的时间，返回格式见下文|function(date)||
|showTime|是否显示时间选择面板|boolean|false|
|showHour|时间选择面板是否显示小时|boolean|true|
|showSecond|时间选择面板是否显示秒|boolean|true|
|timePicker|自己传入定制的 timePicker|React Element|-|
|yearSelectOffset | 年选择器中第一个年份与当前选中值之间的距离，例如当前为 1997 年，距离为 50，则最早可选择年份为 1947 年 | number | 50 |
|yearSelectTotal| 年选择器中年份的总数，如最早可选为 1947 年，总数为 100，则可选年份范围为 1947 - 2047 | number | 100 |
|renderSidebar|sidebar|React Element|() => React.Node
|renderFooter|extra foot|React Element|() => React.Node


#### disabledTime 例子

``` javascript
function range(start, end) {
  const result = [];
  for (let i = start; i < end; i++) {
    result.push(i);
  }
  return result;
}

function disabledTime() {
  return {
    disabledHours: () => range(0, 24).splice(4, 20),
    disabledMinutes: () => range(30, 60),
    disabledSeconds: () => [55, 56],
  };
}
```

### MonthCalendar Props

|参数|说明|类型|默认值|
|---|----|---|------|
|value|日期|string/number|无|
|defaultValue|日期|string/number|无|
|format|展示的日期格式|string|'yyyy-MM'|
|allowedMonthRange|允许的月份范围,与disableDate互斥|object|

### YearCalendar Props

|参数|说明|类型|默认值|
|---|----|---|------|
|value|日期|string/number|无|
|defaultValue|日期|string/number|无|
|format|展示的日期格式|string|'yyyy'|

### RangeCalendar Props

参数基本与 Calendar 相同

|参数|说明|类型|默认值|
|---|----|---|------|
|value|日期|array|无|
|defaultValue|日期|array|无|
|quickSelectRanges|快捷日期选择|array|无|

####  RangeCalendar 例子
``` js
<RangeCalendar
    size="large"
    value={rangeValue}
    quickSelectRanges={[
        {
            text: '19年1、2月份',
            value: {
                start: '2019-01',
                end: '2019-02'
            }
        },
        {
            text: '18年S2',
            value: {
                start: '2018-09',
                end: '2019-03'
            }
        }
    ]}
    onSelect={(v, formatted) => {
      this.onRangeSelect(v, formatted);
    }}
/>
```

### CalendarFull props

|参数|说明|类型|可选值|默认值|
|------|----|---|------|-|
|value|日期|string/number|无||
|defaultValue|默认日期|string/number|||
|type|类型|string|time/week/month|'time'|
|startHour|面板开始时间（小时）|number||9|
|endHour|面板结束时间（小时）|number||23|
|step|面板间隔时间(分钟)|number||60|
|format|展示的日期格式|string|'YYYY-MM-DD'||
|disabledTime|不可选择的时间段，返回格式见下文|function(date)|||
|disabledDate|不可选择的时间日期|function(date)|||
|showHeader|是否显示头部|boolean|true||
|onSelect|当面板日期被选中时触发|function(date)|||
|onTypeChange|当切换面板类型时触发|function()|||
|weekRender|周视图面板中每个格子的渲染函数|function(current,formatDate)|||
|dateRender|月视图面板中每个格子的渲染函数|function(current,formatDate)|||
|timeRender|日视图面板中每个格子的渲染函数|function(current,formatDate)|||
|headerComponent|头部自定义组件|React Element|||
|headerRender|头部自定义渲染函数|function(current,type)|||
|scheduleRender|渲染面板中的函数，可以跨日程显示详情|Calendar.util.generateScheduleContent(events)|||
|getPopupContainer|指定渲染日程的容器|function|||


### MiniWeek props

|参数|说明|类型|可选值|默认值|
|------|----|---|------|-|
|locale|中英文|string|zh-cn/en|zh-cn|
|events|渲染的日程事件|array||[]|
|scheduleRender|用来渲染具体日程下的事件|function(dateInfo)|||

####  CalendarFull 例子
``` js
//events.js
const events =  [
  {
    start: '2018-10-24 13:00', // 事件开始时间
    end: '2018-10-31 23:00', // 事件结束时间
    // 事件的回调函数
    render: ({current,startHour,endHour,step,type}) => {}
  },
];
<CalendarFull
  value={this.state.value}
  onSelect={this.onSelect}
  type="month"
  locale="zh-cn"
  format="yyyy/MM/dd"
  scheduleRender={Calendar.util.generateScheduleContent(events)}
  startHour={12}
  endHour={18}
  step={40}
  />
```

####  MiniWeek 例子
```js
miniWeekRender(dateInfo){
  /**
   * 返回的dateInfo结构如下
   * {
   *   events: [],当前时间下的日程事件
   *   label: 当前时间是星期几
   *   value: 当前日期
   * }
  */
}

<MiniWeek
  locale="zh-cn"
  events={events}
  scheduleRender={this.miniWeekRender}
  >
  <div className="schedule-container">
    <h3>这是日程渲染事件</h3>
  </div
</MiniWeek>
```

```js
 <MiniWeek
  value="2018-12-24"
  locale="en-us"
  events={events}
  scheduleRender={this.miniWeekRender}
  getPopupContainer={() => this.miniRender}
/>
<div className="miniweek-container" ref={(c) => { this.miniRender = c; }} />
```


#### util

> 一些辅助函数，用于某些套餐化定制

* Calendar.util.generateContentRender(code, locale): 用于在日历上标注非常规的休假、上班以及日程。
    * code should be an object like this {'xxxx-xx-xx': ['work/leave/schedule']}
    * locale should be `zh-cn` or `en-us`

* Calendar.util.generateScheduleContent(events):用于在日历中展示跨日程详情
 > Tips:当不确定日程详情中是否有跨日程事件时，推荐使用scheduleRender
   events 为对象数组，格式如下：
```js
[{
    start: '2018-10-16 13:00',//事件开始时间
    end: '2018-10-16 14:00',//事件结束时间
    //事件的回调函数
    render: () => <div>10-10 </div>
}]
```


HISTORY.md
# History

## 0.10.45
* `fix` 增加locale兼容 (#78)

## 0.10.44
* `feat` 支持更完善的国际化方案 (#77)

## 0.10.43
* `feat` 新增 locale key en_US zh_CN (#76)
  
## 0.10.42
* `feat` support localePack

## 0.10.35
* `feat` add new prop onHeaderSelect for calendarfull

## 0.10.34
* `FIXED` fix i18n form RangeCalendar

## 0.10.33
* `CHANGED` rcCalendar support renderFooter

## 0.10.32
* `CHANGED` make a refactor in fullcalendar

## 0.10.31
* `FIXED` fix a style bug in fullcalendar mode when event cross different month panel

## 0.10.29
* `FIXED` fix styles in year panel

## 0.10.28
* `NEW` add events support for vc-calendar

## 0.10.26
* `CHANGED` change some style in calendarfull
* `NEW` add new callback for schedule click

## 0.10.25
* `CHANGED` change date picker offset in MonthCalendar & YearCalendar

## 0.10.24
* `FIXED` locale fix

## 0.10.23
* `FIXED` revert to 0.10.21

## 0.10.22
* `NEW` add new prop firstDayOfWeek

## 0.10.21
* `NEW` Compatible with different locale formats

## 0.10.20
* `FIXED` fix disabledDate & disabledTime is not working when init

## 0.10.19
* `FIXED` placeholder support i18n

## 0.10.18
* `NEW` add miniWeek、fullWeek component and fix some documents


## 0.10.17
* `CHANGE` auto hidden value text when width is too small
           give a min-width to input-span: 80px which can at least show full year like 2019
           fix a wrong prop type of YearCalendar.transitionName

## 0.10.16
* `CHANGE` lock rc-calendar version

## 0.10.15
* `CHANGE` when select a quickSelectRanges, close the tooltip

## 0.10.13
* `FIXED` add new prop quickSelectRanges for RangeCalendar

## 0.10.11

* `CHANGED` modify footer close button text

## 0.10.10

* `NEW` add new prop quickSelectRanges for RangeCalendar

## 0.10.9

* `FIXED` year can not changed when defaultValue is not set in monthCalendar
* `NEW` add new prop allowedMonthRange for MonthCalendar

## 0.10.7

* `FIXED` month & year cannot be changed if defaultOpenValue is set
* `NEW` support FullCalendar thanks to @kwl777

## 0.10.6

* `CHANEGD` `Calendar`, `MonthCalendar`, `YearCalendar`, new prop `defaultOpenValue`
* `FIXED` some eslint error

## 0.10.4

* `FIXED` monthCalendar style bug after updating `rc-calendar` to `^9.6.0`

## 0.10.2

* `CHANGED` support js style export

## 0.10.0

* `CHANGED` update `uxcore-select` to `^0.5.0`
* `CHANGED` update `rc-calendar` to `^9.6.0`

## 0.9.24

* `CHANGED` value can be selected from end to start in cascade mode

## 0.9.23

* `FIXED` less cascade & usage

## 0.9.22

* `CHANGED` update `rc-time-picker` to `^3.0.0`

## 0.9.21

* `CHANGED` update `rc-util` to `4.x`

## 0.9.18

* `FIXED` RangeCalendar select error if value is null or []

## 0.9.17

* `FIXED` RangeCalendar locale presentation error if value is []

## 0.9.16

* `CHNAGED` update `rc-time-picker` to ~2.4.0

## 0.9.15

* `CHNAGED` update `rc-calendar` to ~8.4.0

## 0.9.14

* `CHANGED` support new prop `onOpenChange`

## 0.9.13

* `FIXED` timepicker style bug when calendar is shown around the border of the container.

## 0.9.11

* `NEW` support new prop `allowClear`


## 0.9.10

* `CHANGED` fit React@15

## 0.9.9

* `CHANGED` style adjustment for small & middle size

## 0.9.8

* `CHANGED` support new prop `size`

## 0.9.7

* `CHANGED` support new props `yearSelectOffset` & `yearSelectTotal`

## 0.9.6

* `FIXED` range picker selected color
* `FIXED` range picker width

## 0.9.5

* `FIXED` HeaderSelect can select disabled date.
* `FIXED` HeaderSelect style adaption

## 0.9.4

* `FIXED` add `generalizeFormat` in Month & Year Calendar

## 0.9.3

* `FIXED` incorrect style

## 0.9.2

* `CHANGED` default format for pmam timepicker

## 0.9.1

* `CHANGED` month & year panel style change

## 0.9.0

* `CHANGED` use new icon
* `CHANGED` style details adjustment

## 0.8.5

* `FIXED` fix year select menu will overflow if locale is en-us

## 0.8.4

* `CHANGED` base style change

## 0.8.2

* `NEW` add new sub Component `RangeCalendar`
* `CHANGED` default `yearSelectOffset` & `yearSelectTotal` change

## 0.8.1

* `FIX` fix style bug

## 0.8.0

* `NEW` new style & timePicker

## 0.7.0

* `NEW` new animation

## 0.6.6

* `CHANGED` ESLint: 80%
* `NEW` add new method `getTriggerNode`

## 0.6.5

`FIXED` can not reset input value if value is null in React@15.x

## 0.6.4
`FIXED` text-indent cause scroll issue.

## 0.6.3

`NEW` new props `showSecond` `showHour`

## 0.6.2

`FIXED` missing method `getTimeConfig`

## 0.6.1

`CHANGED` add prop `timezone` to fix calendar timezone

## 0.6.0

`CHANGED` all panel support clear icon

## 0.5.2

`FIXED` fix select time btn style bug

## 0.5.1

`FIXED` fix contentRender bug

## 0.5.0

`CHANGED` update dependency `rc-calendar` to ~6.0.0

## 0.4.3

`NEW` add new prop `inputWidth` for adjusting the trigger width.

## 0.4.2

`NEW` add polish support.

## 0.4.1

* `FIX` fix monthCalendar & yearCander picker input style bug
* `FIX` disable input style bug

## 0.4.0

`FIX` fix trigger icon

## 0.3.9

`FIX` fix style bug in `input[readonly]`

## 0.3.8

`NEW` add prop `getPopupContainer`

## 0.3.7

`NEW` add prop `timePicker`

## 0.3.6

`FIX` fix npm tarball download fail

## 0.3.5

`NEW` add animate effect

## 0.3.3

`NEW` add TransitionName prop

## 0.3.2
`NEW` export CalendarPanel in Calendar

## 0.3.1
`NEW` add util.generateContentRender to provide a custom package for date render.

## 0.3.0
* `CHANGED` new style
* `CHANGED` change rc-calendar to ~5.5.0

## 0.2.6
`FIXED` fix camelcase of picker.js

## 0.2.5
`FIXED` fix issue #12

## 0.2.4
`CHANGED` update scaffold

## 0.2.2

`CHANGED` set timezoneOffset by user's local timezoneOffset

## 0.2.1
`CHANGED` add year calendar

## 0.2.0
`CHANGED` add month calendar

## 0.1.14
`FIXED` #6

## 0.1.13
`FIXED` fix issue #5: calendar cannot be reset by null.

## 0.1.12
`FIXED` use getCalendarContainer to add uxcore wrap

## 0.1.11
`FIXED` sorry, fix https://github.com/uxcore/uxcore-calendar/issues/3 again

## 0.1.9
`FIXED` https://github.com/uxcore/uxcore-calendar/issues/3

## 0.1.7

demo/events.js
const current = +new Date();
export default [
  {
    start: "2019-05-10 14:00", // 事件开始时间
    end: "2019-05-11 16:00", // 事件结束时间
    important: true,
    style: {
      background: "lightblue",
      color: "red",
    },
    // 事件的渲染函数
    render: () => <div>事件1</div>,
  },
  {
    start: "2019-05-10 14:00", // 事件开始时间
    end: "2019-05-14 16:00", // 事件结束时间
    important: true,
    // style: {
    //   background: '#ddd',
    //   color: 'blue'
    // },
    // 事件的渲染函数
    render: () => (
      <div>
        <span>事件2</span>
      </div>
    ),
  },
  {
    start: "2019-05-10 14:00", // 事件开始时间
    end: "2019-05-21 16:00", // 事件结束时间
    important: true,
    style: {
      background: "yellow",
      color: "blue",
    },
    // 事件的渲染函数
    render: () => <div>事件3</div>,
  },
  {
    start: "2019-05-30 14:00", // 事件开始时间
    end: "2019-06-13 17:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件4</div>,
  },
  {
    start: "2019-05-30 14:00", // 事件开始时间
    end: "2019-05-30 18:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件5</div>,
  },
  {
    start: "2019-05-30 14:00", // 事件开始时间
    end: "2019-06-13 21:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件8</div>,
  },
  {
    start: "2019-06-03 14:00", // 事件开始时间
    end: "2019-06-03 21:00", // 事件结束时间
    important: true,
    style: {
      background: "lightblue",
      color: "white",
    },
    // 事件的渲染函数
    render: () => <div>事件9</div>,
  },
  {
    start: "2019-04-29 14:00", // 事件开始时间
    end: "2019-04-30 21:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件10</div>,
  },
  {
    start: "2019-06-05 06:00", // 事件开始时间
    end: "2019-06-06 21:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件11</div>,
  },
  {
    start: "2019-06-05 09:00", // 事件开始时间
    end: "2019-06-08 21:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件12</div>,
  },
  {
    start: "2019-06-05 14:00", // 事件开始时间
    end: "2019-06-05 21:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>事件13</div>,
  },
  {
    start: "2019-06-04 01:00", // 事件开始时间
    end: "2019-06-04 24:00", // 事件结束时间
    important: true,
    // 事件的渲染函数
    render: () => <div>14</div>,
  },
];
